log-viewer/src/main/java/com/logviewer/web/session/LogSession.java

```diff
@@ -334,29 +350,54 @@ public synchronized void cancelSearch() {
 
     @Remote
     public synchronized void searchNext(Position start, boolean backward, int recordCount, SearchPattern pattern,
-                             @NonNull Map<String, String> hashes, long stateVersion, long requestId) {
-        if (this.stateVersion > stateVersion)
+        @NonNull Map<String, String> hashes, long stateVersion, long requestId, boolean loadNext) {
+        if (!checkStateVersion(stateVersion)) {
             return;
+        }
 
-        if (this.stateVersion < stateVersion)
-            throw new IllegalStateException("backend_stateVersion=" + this.stateVersion + ", but UI_stateVersion=" + stateVersion);
-
-        CompletableFuture<SearchTask.SearchResponse> execution = execute(new SearchTask(sender, logs, start, recordCount, backward, pattern, hashes, filter));
-
-        execution.whenComplete(new LogExecutionHandler<SearchTask.SearchResponse>() {
-            @Override
-            protected void handle(SearchTask.SearchResponse res) {
-                sender.send(new EventSearchResponse(res.getStatuses(), stateVersion, res, requestId));
+        SearchTask searchTask = new SearchTask(sender, logs, start, recordCount, backward, pattern, hashes, filter);
 
-                if (res.getData() != null) { // if found
-                    int foundIdx = backward ? 0 : res.getData().size() - 1;
-                    Record found = res.getData().get(foundIdx).getFirst();
-                    assert pattern.matcher().test(found.getMessage());
+        execute(searchTask).thenCompose(searchRes -> {
+            if (searchRes.getData() == null) {
+                return CompletableFuture.completedFuture(new EventSearchResponse(searchRes, stateVersion, requestId));
+            }
 
-                    loadNext(new Position(found, backward), backward, recordCount, hashes, stateVersion);
+             
+            int foundIdx = backward ? 0 : searchRes.getData().size() - 1;
+            Record found = searchRes.getData().get(foundIdx).getFirst();
+            assert pattern.matcher().test(found.getMessage());
+            
+            if(loadNext) {
+                LoadRecordTask loadTask = new LoadRecordTask(sender, logs, recordCount, filter, new Position(found,
+                    backward), backward, hashes);
+                try {
+                    LoadNextResponse loadRes = execute(loadTask).get(taskTimeoutMS, TimeUnit.MILLISECONDS);
```

Comment:
if the timeout occurs, the result of `execute(loadTask)` will be never send to UI, although we spent resources to compute that.